homeassistant/components/tradfri/base_class.py

```diff
@@ -44,103 +36,37 @@ async def wrapper(command: Command | list[Command]) -> None:
-            None
-        )
-        self._device_data: Socket | Light | Blind | AirPurifier | None = None
-        self._gateway_id = gateway_id
-
-    async def _async_run_observe(self, cmd: Command) -> None:
-        """Run observe in a coroutine."""
-        try:
-            await self._api(cmd)
-        except PytradfriError as err:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
-
-    @callback
-    def _async_start_observe(self, exc: Exception | None = None) -> None:
-        """Start observation of device."""
-        if exc:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._attr_name, exc_info=exc)
-        cmd = self._device.observe(
-            callback=self._observe_update,
-            err_callback=self._async_start_observe,
-            duration=0,
-        )
-        self.hass.async_create_task(self._async_run_observe(cmd))
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        self._async_start_observe()
-
-    @callback
-    def _observe_update(self, device: Device) -> None:
-        """Receive new state data for this device."""
-        self._refresh(device)
+        super().__init__(device_coordinator)
 
-    def _refresh(self, device: Device, write_ha: bool = True) -> None:
-        """Refresh the device data."""
-        self._device = device
-        self._attr_name = device.name
-        if write_ha:
-            self.async_write_ha_state()
+        self._gateway_id = gateway_id
 
+        self._device: Device = device_coordinator.data
 
-class TradfriBaseDevice(TradfriBaseClass):
-    """Base class for a TRADFRI device.
+        self._device_id = self._device.id
+        self._api = handle_error(api)
+        self._attr_name = self._device.name
 
-    All devices should inherit from this class.
-    """
-
-    def __init__(
-        self,
-        device: Device,
-        api: Callable[[Command | list[Command]], Any],
-        gateway_id: str,
-    ) -> None:
-        """Initialize a device."""
-        self._attr_available = device.reachable
-        self._hub_available = True
-        super().__init__(device, api, gateway_id)
+        self._attr_unique_id = f"{self._gateway_id}-{self._device.id}"
 
     async def async_added_to_hass(self) -> None:
         """Start thread when added to hass."""
-        # Only devices shall receive SIGNAL_GW
-        self.async_on_remove(
-            async_dispatcher_connect(self.hass, SIGNAL_GW, self.set_hub_available)
+        self.async_on_remove(  # Only devices shall receive SIGNAL_GW
+            async_dispatcher_connect(
+                self.hass, SIGNAL_GW, self.coordinator.set_hub_available  # type: ignore
```

Comment:
On this line, I'm getting the following:
```
homeassistant/components/tradfri/base_class.py:65: error: "DataUpdateCoordinator[Any]" has no attribute "set_hub_available"  [attr-defined]
Found 1 error in 1 file (checked 2 source files)
```

What is the correct way to go about here? Looks like `self.coordinator` only references `DataUpdateCoordinator`?

Reply:
I think we should put the callback on the coordinator and not on the entity.

Reply:
Maybe add a setup method to the coordinator where we can connect the callback?

Reply:
Regarding the typing of the coordinator I've added some suggestions below.

Reply:
Thank you!